To handle cluster connection fail in GUI

twentyse7en opened the pull request:
fix issue #601 
## TODO
* [x] Handle connection failure in server
* [x] Display error message
* [x] <strike> Reset submit button</strike> Show spinner while connecting to server
 
## Contributor Checklist:

* [x] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [x] I have added/updated test cases

sk1p commented:
Closing and re-opening PR to try and fix travis status

sk1p commented on src/libertem/web/connect.py:
```diff
@@ -39,9 +39,13 @@ def initialize(self, state: SharedState, event_registry):
+                ))
+                connected = True
+            except OSError:
+                connected = False
```
The `connected` approach here works, but makes it a bit hard to follow the execution of the code. What do you think about just writing the failure response here, and returning? It also means you keep the context of the exception when returning the error response (see below)

twentyse7en replied:
Yep! that would be very clean and clear.

sk1p commented on src/libertem/web/connect.py:
```diff
@@ -57,23 +61,31 @@ def initialize(self, state: SharedState, event_registry):
+                "connection": connection,
+            })
+        else:
+            self.write({
+                "status": "disconnected",
+                "connection": {},
+            })
```
This could (should?) also return the concrete error message from the exception - possibly also using `"status": "error"` instead? See also `messages.py`, where several `_error` messages are defined. If you like, you can also extract the response into a method of the `Messages` class, so it won't be forgotten once the schema project is implemented.

twentyse7en commented:
>
>The GUI properly goes back to the "not connected" status after an error, right? 

Yes
>
>Do you want to add error displaying functionality to the GUI, or should that be done as a follow-up PR?

Will do it here, If there are no other problems : )

sk1p commented:
> Will do it here, If there are no other problems : )

:+1:

sk1p commented on tests/server/test_startup.py:
```diff
@@ -93,6 +93,21 @@ def _get_raw_params(path):
         }
 
 
+@pytest.mark.asyncio
+async def test_cluster_connect_error(base_url, http_client):
+    url = "{}/api/config/connection/".format(base_url)
+    conn_details = {
+        'connection': {
+            'type': 'TCP',
+            'address': 'tcp://unknown:8786',
+        }
+    }
```
:+1: - instead of "unknown", you could also [use the "invalid" name that is guaranteed to not resolve to any hostname](https://tools.ietf.org/html/rfc6761#section-6.4) (RFC 6761). It should also guarantee that the name resolution fails early, so the error should happen earlier ("Name resolution APIs and libraries SHOULD recognize "invalid" names as special and SHOULD always return immediate negative responses.  Name resolution APIs SHOULD NOT send queries for "invalid" names to their configured caching DNS server(s).")

Maybe this will also fix [the timeout issue here](https://travis-ci.org/github/LiberTEM/LiberTEM/jobs/670861793?utm_medium=notification&utm_source=github_status)?

twentyse7en replied:
"invalid" address is raising a `ValueError: missing port number in address 'invalid'`. It only takes around 8 ms to respond. For "invalid:8080" it takes 10s to respond.

sk1p replied:
Very nice - yeah I think we can't get around the 10s timeout without doing something funky (mocking the `Executor` comes to mind...) - I guess that's a retry-loop somewhere in Dask that retries even the DNS lookup :grinning: 

I like the new solution, let's keep it :+1:

twentyse7en commented:
How can I display the error message, there is no action for error in the cluster.

sk1p commented:
> How can I display the error message, there is no action for error in the cluster.

Yeah, error displaying implementation in the GUI is a bit rough, it needs a refactor I think. For now, you can go ahead and add a "cluster error" action, add that to the `errorReducer` and be done with it.

twentyse7en commented:
I'm stuck :crying_cat_face: , what to do next!

sk1p commented:
> I'm stuck, what to do next!

One thing to always do is to keep the CI pipeline happy and green. In this case, the [typescript build and lint job](https://travis-ci.org/github/LiberTEM/LiberTEM/jobs/672114639) is easily fixed I think - not sure what the problem with [the other job is](https://travis-ci.org/github/LiberTEM/LiberTEM/jobs/672114631) - possibly an intermittent problem? (For TypeScript: see the "Organize Imports" command of vs code, if you are using that!)

Then, if you are stuck, sometimes it is good to take a step back and have a look at the original problem again. What were you trying to accomplish? What were the parts of the whole change, and which parts are already done? What is missing? If you trigger the new behavior, what is happening, what should happen? Maybe also re-read some of the previous discussion and issues. This is most of the time enough for me to get unstuck. If you still need guidance then, please let me know!

I also like to make a checklist in the process, which can then remind me later what is still to be done.

twentyse7en commented:
Resolved the lint error. In the previous commit forget to build the client :sweat_smile: how can I resolve conflict?

I have added `clusterError` action, also added to `errorReducer` and dispatch action when the connection fails in `sagas.ts`. But the error message isn't popping up : /

sk1p commented:
> Resolved the lint error

:+1:

> I have added `clusterError` action, also added to `errorReducer` and dispatch action when the connection fails in `sagas.ts`. But the error message isn't popping up : /

Hmm. Have you verified that the action is dispatched? This can be done with the redux dev tools. There you can also have a look at the action and see if it contains the data you expect it to contain. You can also inspect the complete redux state before/after the action is dispatched. Let me know if this helps!

> how can I resolve conflict?

Do you know how this is usually done, when you get a conflict in a rebase? The issue here is that the conflicts are in generated code - so it would be hard to manually resolve this. What I usually do is I start to rebase to master as usually, but then when there are conflicts in those `web/client/**` files, I just rebuild the client again, which should make sure that your changes take effect. Then I do `git rebase --continue` as usual...

I'm sure there is a better way, but that's how I do it :grinning: 

If you want to practice this, I would highly suggest to just create a new branch from the current one as backup, so if you mess up, you can just go back to the old branch.

twentyse7en commented:
![error_handle](https://user-images.githubusercontent.com/59721339/78794187-0bb8af00-79d1-11ea-9054-55d448124334.png)

Action is dispatched, where it should receive this information and display the message? I'm a beginner in react, can you go through the code and give the necessary suggestions. I did try different ways but nothing works.

I have never resolved conflicts. Thanks for the suggestion : )

sk1p commented:
> where it should receive this information and display the message?

This, or similar, is how an error message is supposed to look like:

![image](https://user-images.githubusercontent.com/5778/78796857-35fd7300-79b7-11ea-84f3-154797e40318.png)

(You can easily trigger this one yourself by entering a non-existing path into the file browser, to compare with what happens with your own error.)

Then, you can select your `CLUSTER_ERROR` action and check the "State" tab. If you expand the `errors` key there, you should see your error message:

![image](https://user-images.githubusercontent.com/5778/78797199-ac9a7080-79b7-11ea-9366-0f814165bfb7.png)

The "Diff" tab should also show that the `errors` key changed. If this is not the case, maybe your change in the reducer doesn't work for some reason?

Let me know if this helps, if you are still stuck I can check out your branch and try to help debug this.

twentyse7en commented:
![Screenshot from 2020-04-09 13-03-05](https://user-images.githubusercontent.com/59721339/78888605-97891480-7a7f-11ea-8bd3-9dca97211f21.png)

Here I tried to connect to the cluster first, so it dispatched a `cluster error`. Sadly no error message popped up. Then connected to local cluster. It was successful, now the previous error message is showing.I'm still stuck can you check my branch?

sk1p commented:
Hm, I'll have a detailed look later, but: is it possible that the error messages are not even shown before the cluster connection is there? Hint: have a look at `ErrorList.tsx`! That may need changes...

twentyse7en commented:
In `ErrorList.tsx`

`
const mapStateToProps = (state: RootReducer) => {
    return {
        errors: state.errors,
        clusterConnected: (state.clusterConnection.status === "connected" ||
            state.clusterConnection.status === "disconnected"),
        channelConnected: (state.channelStatus.status === "connected" ||
            state.channelStatus.status === "ready"),
    }
}
`
Now it's working!! Is that ok or do we need any other changes?

sk1p commented:
> `clusterConnected: (state.clusterConnection.status === "connected" || state.clusterConnection.status === "disconnected")`

Hm, that's a bit misleading, as `clusterConnected` will be `true`, even though the cluster connection status is `"disconnected"`. Maybe you can have a look where the variable is used, and change that instead?

twentyse7en commented:
Yeah! why was there a constrain to show error only after the cluster connection is successful ?

sk1p commented:
> Yeah! why was there a constrain to show error only after the cluster connection is successful ?

I'm not sure - possibly in an older version there were spurious errors in that state. Did you get any other errors when testing? You can try to remove that constraint - if any spurious errors show up, we can fix those in a different way.

twentyse7en commented:
I removed the constrain and its working fine : )

sk1p commented:
For the last point on your list, a small hint: https://jaredpalmer.com/formik/docs/1.5.8/guides/form-submission

(note that we are still using an earlier version of formik, as we have not migrated to 2.x yet)

twentyse7en commented:
Thx :+1: So when the server responds with an error, by calling `setSubmitting(false)` the form can be reset. We are making the request in `api.ts`, is this the place to handle this?

sk1p commented:
> Thx  So when the server responds with an error, by calling `setSubmitting(false)` the form can be reset. We are making the request in `api.ts`, is this the place to handle this?

Hmm, looking at it again: `setSubmitting(false)` is only available in the form itself and `handleSubmit` function etc. -  it can't easily be called from a saga or from inside the API helper function. (Aside: API helper functions like `connectToCluster` are supposed to do only one thing, and that is to speak with the HTTP API. Ideally other things, like interacting with forms, would be done by the callers of those functions.)

Maybe it is easier to introduce a new cluster status `"connecting"`, replace the whole form with a spinner (like the `channel` status is handled - you can test this by stopping the libertem server but keeping the browser tab open - it should show the spinner and try to reconect.)

An alternative would be to completely take the cluster state out of redux/redux-saga and into a react component. I think this may be harder for you to do, because this would change the way the state is handled completely.

twentyse7en commented:
I have no idea why tests are failing! @sk1p can you have a look

sk1p commented:
> I have no idea why tests are failing! @sk1p can you have a look

Sorry, yeah, these appear to be unrelated to your pull request. I've opened a separate issue to track this: https://github.com/LiberTEM/LiberTEM/issues/713

twentyse7en commented:
>
>If a user runs into this bug, they probably want to re-try with similar parameters, for example if they had a typo in the URL, or the remote cluster was not running yet. So it would be really nice if the form re-opened with the original parameters.

Sounds good! I'll try to implement that :+1:
